### ISSUE TITLE
Showing existing, or maybe generating thumbnails for MP4 and WEBM

### ISSUE THREAD

**Issue by deterenkelt** [BestImageViewer/geeqie#541]
### ISSUE TYPE

* Feature request

### GEEQIE VERSION

```
Geeqie b0ea0793
```

### SUMMARY

Geeqie shows generic icons for videofiles, but it would be better to show real thumbnails. There is one simple program [ffmpegthumbnailer](https://github.com/dirkvdb/ffmpegthumbnailer), that Geeqie could just pull as a dependency and call as a binary. 

I’ve read in #510, that we have no thumbnails, because there would be a lot of dependencies (to GStreamer for example). This one depends on ffmpeg and libjpeg.

Also Geeqie could at least try to load thumbnails for videos, there is usually a thumbnailing daemon like tumbler in XFCE. I use it to generate thumbnails outside the folders, that I browse with Geeqie, and for video files everywhere. So there’s enough thumbnails already pre-generated.

**Comment by tomaszg7** [comment]
I considered using https://github.com/dirkvdb/ffmpegthumbnailer before. Turns out it is not that difficult. Although there is no documentation and library part is quite limited. 

I created an experimental patch adding support for both thumbnails and main image preview of video files. I enabled an option in ffmpegthumbnailer to overlay film strip to distinguish them from images (although that would make sense for both photos and movies ;) ). 

I'm not sure if I did it correctly, but it seems to work for me. For some reason calling `area_prepared_cb` breaks it. Maybe it is some kind of race condition, I don't know. I'm not also sure if I need to do scaling in image loader backend or is it OK to return larger pixbuf. Regretfully ffmpegthumbnailer accepts only `size` dimension, not width and height separately.

[geeqie-video-thumbnails.patch.gz](https://github.com/BestImageViewer/geeqie/files/1464017/geeqie-video-thumbnails.patch.gz)

**Comment by tomaszg7** [comment]
Also, the size of overlayed film strip is determined by ffmpegthumbnailer itself, and for movies in portrait orientation it is hardly visible on the thumbnail.

Moreover, thumbnails for some landscape movies turn out upside down.

**Comment by deterenkelt** [comment]
It works! Thank you very much!

**Comment by caclark** [comment]
Hello Tomasz
With the few clips I have this works well.

However I get compile errors:
I only have libffmpegthumbnailer 2.1.1 on my system. This seems to have been around for a while, and works (for me, at least).
The other dependencies are
libpng and libjpeg.
The -dev versions of all three are required.

The -dev packages are not checked for anywhere in configure.in, and yet without them you get compile errors. I don't know why they are not checked for.

**Comment by tomaszg7** [comment]
I tested only with libffmpegthumbnailer 2.2.0 (and that's the version I put into configure.in). I'll test earlier versions and look into lowering that requirement.

> The -dev packages are not checked for anywhere in configure.in,
>  and yet without them you get compile errors. I don't know why they are not checked for.
I guess that it is a bug in libffmpegthumbnailer PKGCONFIG file. It should include `Requires` entries. It adds e.g. linker flags to jpeg and png by hand. I'm not an expert but it seems ugly.

However I think the dependence on jpeg headers was a bug in my code, I didn't edit `#ifdef` in `image_load_ffmpegthumbnailer.h`. I don't see the dependence on the level of ffmpegthumbnailer headers. There is also a runtime dependency on libpng - i.e. pixbuf should be able to load it.

**Comment by tomaszg7** [comment]
> I tested only with libffmpegthumbnailer 2.2.0 (and that's the version I put into configure.in). I'll test earlier versions and look into lowering that requirement.

Regretfully I'm unable to build versions 2.0.0 or 2.1.0 since my FFmpeg is to recent. I think that if you tested that it works on 2.1.x, we could put that as a minimal version. 2.0.x are probably to ancient to bother. However using 2.2.0 adds the possibility to try to use embedded image in the stream (but I didn't explore that possibility yet).

**Comment by caclark** [comment]
> version. 2.0.x are probably to ancient to bother. 

I'm using Ubuntu 17.04, which expires end of 2017. But I won't be able to upgrade until mid-December.
I guess other people will have already done so by then and will have 2.2.0 available.

The Ubuntu LTS versions all use version 2.0.x of libffmpegthumbnailer, so I guess they are out of the question anyway.
So, to specify 2.1.1 just for Nov/Dec? Probably not worth it.

But that's just Ubuntu-world.

**Comment by tomaszg7** [comment]
I don't know if 2.0.x are out of the question, API looks (almost) the same. Adding `AC_CHECK_MEMBER` to deal with metadata support, it should be possible to make it work with 2.0.x. It is probably worthwhile to target as big chunk of systems, and Ubuntu-world is the biggest :)

I'm attaching the updated patch and I'll try to test if it really works with 2.0.x
[geeqie-video-thumbnails.2.patch.gz](https://github.com/BestImageViewer/geeqie/files/1464767/geeqie-video-thumbnails.2.patch.gz)

**Comment by caclark** [comment]
Hello Tomasz
I think there is a small problem when compiling with --disable-ff...
You put someone else as the author of one of the new files.

**Comment by tomaszg7** [comment]
Fixed the problem with --without-ff.... One missing ifdef. I also added support for receiving messages from the library. Unfortunately, it crashes on one file I stumbled upon (https://github.com/dirkvdb/ffmpegthumbnailer/issues/162).

Moreover, the problem with rotating some videos is already fixed in git.

[geeqie-video-thumbnails.3.patch.gz](https://github.com/BestImageViewer/geeqie/files/1465185/geeqie-video-thumbnails.3.patch.gz)

**Comment by tomaszg7** [comment]
> You put someone else as the author of one of the new files.

You refer to the fact that I put myself in .c file (I don't know if I was supposed to, checked ..._jpeg.c file) or that I left the original author in .h file (it is just a copy of other files, I didn't do anything there)?

**Comment by caclark** [comment]
image_load_ffmpegthumbnailer.h. You're the author of this new function. You should be taking the credit for all the additions on this :-)

**Comment by caclark** [comment]
Looks good. Do you want this to go into the repository now, or do you want to wait for those other problems to be resolved?

**Comment by tomaszg7** [comment]
Maybe let's wait a few days. The crash thing just got resolved in git HEAD. 

Also, maybe there should be some options to change the behaviour of the code. Some that come to mind are:
* enable/disable thumbnail generation
* enable/disable film strip overlay
* enable/disable use of embedded metadata
* seek % (point from where the frame is grabbed)

Also, I was a bit uncertain what to do with returned name and mime types, however they don't seem to be extensively used, so I think it doesn't matter in this case. But maybe some speedup can be obtained by enabling generation of pre-scaled thumbs, like in jpeg code. I also have to check if scaling in my code is necessary in this case.

**Comment by caclark** [comment]
Ok.
Here's my guess at the document changes - 
[ffmpeg-docs.patch.gz](https://github.com/BestImageViewer/geeqie/files/1466937/ffmpeg-docs.patch.gz)

**Comment by tomaszg7** [comment]
Thanks, I'll add it to the next patch. I would change "thumbnail" to "preview and thumbnail" as the frame is displayed also in main window. Moreover, I would remove "digital camera" as it works for all video clips.

Also, is it necessary to mention libpng and libjpg explicitly? Aren't they already needed by GTK/gdk-pixbuf? Even so, I don't think it is useful to list all dependencies of a given library. That goes without saying that they all have their requirements that need to be satisfied and unless they are exotic, I don't see a reason to list them. Moreover I hope that after taking into account https://github.com/dirkvdb/ffmpegthumbnailer/issues/101, it might not be necessary anymore.

BTW. Looking at README.md and install guides I notice there are no references to extra pixbuf-loaders in optional libraries section. Also, `markdown` should also be needed to build from git, and released version tarballs should rather contain those files pre-generated. (Is it possible to run `gen_*` scripts from `autogen.sh` rather than `make`? On the other hand, it might not get updated often enough...)

**Comment by mzagrabe** [comment]
@caclark ,

Do you know if the 1.4 release will have the thumbnailing? I'm not sure if there is branch or just the patches floating in this bug report.

Thanks!

-m

**Comment by tomaszg7** [comment]
Patches are only 1 day old and under heavy development. So for now they are only in this issue. However if you wish, I can create a branch in my forked repo.

**Comment by mzagrabe** [comment]
Hey Tomasz,

Thanks for the offer of a branch. I don't need it. I was just skimming the
geeqie page looking for release info and came across the thumbnail thread.
Very nice!

Thanks for contributing to geeqie!

-m

**Comment by tomaszg7** [comment]
I've been checking some resizing things and I start to believe something is wrong in the code. I couldn't get Geeqie to request a scaled down thumbnail and from my check even for jpg it doesn't do that (disabled thumbnails to test). It always requests size `h=0,w=0`. Seems that `load` part and `set_size` are called in the wrong order. Even in principle something is fishy. In order to compute requested size, Geeqie has to already know original image size. And it knows that only after it gets loaded. 

I run Geeqie in a directory containing one movie and one jpg file. I got in the log:

```
FT requested size w=0:h=0 for /tmp/20170114_134325.mp4
FT requested size w=0:h=0 for /tmp/20170114_134325.mp4
computing size, w=1280, h=720, rw=128, rh=96
setting size, w=128, h=72
JPG requested size w=0:h=0
computing size, w=312, h=300, rw=128, rh=96
setting size, w=99, h=96
JPG requested size w=0:h=0
```
So, each picture is decoded twice and in both cases full sized. Computed size never gets used. I don't know why in the case of JPG the second call to load a picture comes only after computing size, and in the case of FT, both calls to load happen before computing. Maybe it is a race. Moreover JPG loader code later redefines requested_[width,height] values to match size of the actual picture (one of which is always wrong `w=1`) and ignores it afterwards. Either I don't understand how it works, or something is wrong. The first choice is more probable ;) but the comparison of behaviour of GDK loader and JPEG/TIFF suggests otherwise :/ Also there is a comment in `thumb.c`:
```
/* this will speed up jpegs by up to 3x in some cases */
```

The fact that each image is decoded twice is just result of the fact that both thumbnail and full-size version are needed. It may not be optimal, but it is not a huge waste.

I got into all this mess trying to use updated size-related part of FT API. However it seems that for now I would probably comment it all out. But one good thing came out of it since I realized that I really didn't need an extra backend function after all. So the code got simpler. 

Anyhow, I'm attaching updated version with some configure tests for the new and shiny API parts, docs update and so on. It is also heavily sprinkled with debug calls, which I will remove when I figure out this scaling thing.

[geeqie-video-thumbnails.4.patch.gz](https://github.com/BestImageViewer/geeqie/files/1469255/geeqie-video-thumbnails.4.patch.gz)

**Comment by tomaszg7** [comment]
I think I have a version that can be committed. I didn't test it with older ffmpegthumbnailer, but I don't think many people will be using new Geeqie with old libraries ;) Anyway, that's what bug reports are for ;) For reference, there is a nice website to track those things: https://repology.org/metapackage/ffmpegthumbnailer/versions.  I introduced a preferences option to control whether embedded metadata should be used.

I still think that the prescaling code in image-load related parts of Geeqie is at least flawed and probably doesn't work at all. I'll try to analyse that separately from FT thing.

I'm also attaching two separate patches - one to include a bit more extensions in default video filters, and the other to remove some part of dead code that was floating around. Probably when the scaled loading thing gets sorted out, even more code could be removed.

[geeqie-image-load-dead-code.patch.gz](https://github.com/BestImageViewer/geeqie/files/1488508/geeqie-image-load-dead-code.patch.gz)
[geeqie-video-filefilter-defaults.patch.gz](https://github.com/BestImageViewer/geeqie/files/1488509/geeqie-video-filefilter-defaults.patch.gz)
[geeqie-video-thumbnails.5.patch.gz](https://github.com/BestImageViewer/geeqie/files/1488510/geeqie-video-thumbnails.5.patch.gz)

**Comment by caclark** [comment]
Hello Tomasz

I get this compile error from the latest patch:
`  CC       img-view.o
  CC       jpeg_parser.o
image_load_ffmpegthumbnailer.c:109:8: error: member reference base type 'void' is not a structure or union
        loader->vt->thumbnail_size = MAX(lft->requested_width,lft->requested_width);
        ~~~~~~^ ~~
1 error generated.
`

The error pointed to is loader->vt

I am still using Ubuntu 17.04.
Autogen produces:

> Settings:
  Developer:     no
  Debug flags:   no
  Debug log:     yes
  Deprecated:    no
Support:
  LCMS:          yes
  Exiv2:	 yes
  Lirc:          no
  Clutter:       yes
  Libchamplain:		yes
  Libchamplain-gtk:	yes
  Lua:	         yes
  FFmpegthumbnailer:	yes
Documentation:
  Doxygen:       
  doc-tool:      /usr/bin/gnome-doc-tool

Please let me know what other data you need.

**Comment by tomaszg7** [comment]
Aggh, I fixed it everywhere but I missed one spot where it was hidden behind ifdef. `loader->vt` should be replaced by `lft->vt`. Here's the updated file.

[geeqie-video-thumbnails.5.patch.gz](https://github.com/BestImageViewer/geeqie/files/1495840/geeqie-video-thumbnails.5.patch.gz)

**Comment by caclark** [comment]
Implemented in commit 978ec2dd at geeqie.org
